### DISCUSSION TITLE
SQLAlchemy multi-table mapping insert on attribute change

### DISCUSSION THREAD

**Discussion by ljmc-github** [sqlalchemy/sqlalchemy#8034] [Usage Questions] [answered]
I have an entity called a Report which points to a some report stored in a repository, and some repositories can have versioning, therefore the Report entity has an optional version. 

I am now trying to track the reports in SQL via SQLalchemy. Because one report title can have several versions I wanted to [map the entity over two tables](https://docs.sqlalchemy.org/en/14/orm/nonstandard_mappings.html#mapping-a-class-against-multiple-tables), `report` for the report identification (title and other data I do not include here for simplicity) and `report_version` for report versions which reference a report (again plus other data not included here).

I have managed all the above in the following code. But now, I am blocked by the fact that setting the `Report.version` to a newer version causes an update of `report_version` (see `r1` below). I would like it to instead trigger the insertion of a new version in `report_version` (see `r2` below). 

```py
from dataclasses import dataclass, field

from sqlalchemy import Column, ForeignKey, Integer, String, Table, \
    create_engine, join, select, text
from sqlalchemy.orm import Session, column_property, registry

mapper_registry = registry()

@dataclass
class Report:
    title: str
    version: str | None = field(default=None)  # versioning is optional

report = Table(
    "report",
    mapper_registry.metadata,
    Column("report_pk", Integer, primary_key=True),
    Column("title", String(35), nullable=False),
)

report_version = Table(
    "report_version",
    mapper_registry.metadata,
    Column("version_pk", Integer, primary_key=True),
    Column("report_fk", ForeignKey("report.report_pk"), nullable=False),
    Column("version_id", String(1024), nullable=True),
)

mapper_registry.map_imperatively(
    Report,
    report.join(report_version),
    properties={
        "id": column_property(report.c.report_pk, report_version.c.report_fk),
        "version": report_version.c.version_id,
    },
)

engine = create_engine("sqlite://", echo=True, future=True)

mapper_registry.metadata.create_all(engine)

session = Session(engine)

# setting attribute causes ORM to update `report_update`
r1 = Report(title="r1", version="a")
session.add(r1)
session.flush()  # flush r1a
r1.version = "b"
session.flush()  # flush r1b, updates report_version a to b

# this is what I would like to achieve
r2 = Report(title="r2", version=".1")
session.add(r2)
session.flush()  # flush r2.1
session.execute(
    text(
        "INSERT INTO report_version (report_fk, version_id) VALUES (:report_id, '.2')"
    ),
    {"report_id": r2.id},
)
session.flush()  # flush r2.2, adds report_version .2

session.execute(text("SELECT * FROM report")).all()  # r1 and r2
session.execute(text("SELECT * FROM report_version")).all()  # b, .1 and .2

session.execute(select(Report)).scalars().all()  # r1b, r2.1 and r2.2

session.close()
```

**Comment by zzzeek** [comment, marked as answer]
we have two examples for versioning rows at https://docs.sqlalchemy.org/en/14/orm/examples.html#versioning-objects.   one does an inline conversion of an UPDATE into an INSERT, the other has a separate history table.    if you are looking for a "no UPDATE approach" (it's not clear what you want between the two tables), you can look at "versioning using temporal rows".  

if you wanted to split UPDATE / INSERT between the two tables on that join while having them mapped together as you have (and not in two separate ORM classes linked with relationship()), that's not so easy.  I'd take a look at the examples given and see if they can work for you.

**Reply by ljmc-github** [reply]
What I'm trying to achieve sounds like the versioning in history table example you linked:
A report named "how to version in SQLAlchemy ?" will be stored in `report` table and it's versions "a", "b", "rc1", "rc2", "1.0.0" in `report_version`. I'll have a further look into that example and thank you very much.
